src/detours.cpp

```diff
@@ -333,14 +351,67 @@ struct _DETOUR_TRAMPOLINE
     _DETOUR_ALIGN   rAlign[8];      // instruction alignment array.
     PBYTE           pbRemain;       // first instruction after moved code. [free list]
     PBYTE           pbDetour;       // first instruction of detour function.
-    BYTE            rbCodeIn[8];    // jmp [pbDetour]
+    BYTE            rbCodeIn[8];    // jmp [pbDetour]
+    BYTE            rbContext[53];  // context callback code (calls context function w/ void* param)
+    PBYTE           pbContext;      // first instruction of context callback code (rbSetContext) if used
+    PF_DETOUR_INVOKE_CONTEXT_CALLBACK pfContextCallback;    // context callback used as indirect jmp in rbContext
 };
 
-C_ASSERT(sizeof(_DETOUR_TRAMPOLINE) == 96);
+C_ASSERT(sizeof(_DETOUR_TRAMPOLINE) == 168);
 
 enum {
     SIZE_OF_JMP = 5
-};
+};
+
+inline PBYTE detour_gen_context_callback(PBYTE pbCode, PF_DETOUR_INVOKE_CONTEXT_CALLBACK* ppfContextCallback, PVOID pContextParam)
+{
+    PBYTE pbJmpSrc = pbCode + 37;
+
+    // Backup volatile registers that are used to pass __fastcall params.
+    // If the set context function performed floating point operations, 
+    // registers XMM0, XMM1, XMM2, and XMM3 would also have to be backed 
+    // up and restored. In the attempt to keep this code generation as simple 
+    // as possible, floating point operations in the set context function isn't 
+    // currently supported.
+    *pbCode++ = 0x51;   // push rcx (backup rcx)
+    *pbCode++ = 0x52;   // push rdx (backup rdx)
+    *pbCode++ = 0x41;   // push r8 (backup r8)
+    *pbCode++ = 0x50;
+    *pbCode++ = 0x41;   // push r9 (backup r9)
+    *pbCode++ = 0x51;
+
+    *pbCode++ = 0x48;   // sub rsp,20h (required by x64 __fastcall)
+    *pbCode++ = 0x83;
+    *pbCode++ = 0xEC;
+    *pbCode++ = 0x20;
+
+    *pbCode++ = 0x48;   // mov rcx, imm64 (return address)
+    *pbCode++ = 0xB9;
+    *((INT64*&)pbCode)++ = (INT64)(pbJmpSrc);
+    *pbCode++ = 0x51;   // push rcx (return address)

```

Comment:
Why push return address instead of call below?

Reply:
I've ported it as close to the original as I could do, while updating it for style.  I'm open to suggestions on changing it, but I can't really give any reasons on why it was done the way it is.